File: CS280Project1.py
## Abhinav Bassi
## 31327046
## CS 280 102
## Project 1

def othertower(first, second):
    towers = [first, second]
    if "A" not in towers:
        return "A"
    elif "B" not in towers:
        return "B"
    else:
        return "C"

def towerofhanoi(start, end):
    n = len(start)
    if n == 0:
        pass
    else:
        if start[0] == end[0]:
            towerofhanoi(start[1:], end[1:])
        else:
            temp = othertower(start[0], end[0])
            towerofhanoi(start[1:], temp*(n-1)) 
            print ("Move disk from ", start[0], " to ", end[0]) 
            towerofhanoi(temp*(n-1), end[1:])

def main():
    start = input("Enter start configuration: ")
    end = input("Enter end configuration: ")
    towerofhanoi(start, end)

main()

File: CS280Project1.cpp
// Abhinav Bassi
// 31327046
// CS 280 102
// Project 1

#include <iostream>
#include <list>

using namespace std;

char othertower(char first, char second) {
    list<char> towers;
    towers.push_back(first);
    towers.push_back(second);
    if (std::find(towers.begin(), towers.end(), 'A') == towers.end()) {
        return 'A';
    }
    else if (std::find(towers.begin(), towers.end(), 'B') == towers.end()) {
        return 'B';
    }
    else {
        return 'C';
    }
}

string generatetemp(char original, int numrepeat) {
    string tempstring = "";
    for (int i = 0; i < numrepeat; i++) {
        tempstring += original;
    }
    return tempstring;
}

void towerofhanoi(string start, string end) {
    int n = (int) start.length();
    if (n != 0) {
        if (start[0] == end[0]) {
            towerofhanoi(start.substr(1), end.substr(1));
        }
        else {
            string temp = generatetemp(othertower(start[0], end[0]), n-1);
            towerofhanoi(start.substr(1), temp);
            cout << "Move disk from " << start[0] << " to " << end[0] << endl;
            towerofhanoi(temp, end.substr(1));
        }
    }
}

int main(int argc, char *argv[]) {
    string start;
    cout << "Enter start configuration: ";
    cin >> start;
    string end;
    cout << "Enter end configuration: ";
    cin >> end;
    towerofhanoi(start, end);
    return 0;
}
